conbench/api/_examples.py

```diff
@@ -127,7 +127,7 @@ def _api_benchmark_entity(
         "tags": {
             "id": case_id,
             "compression": "snappy",
-            "cpu_count": 2,
+            "cpu_count": "2",
```

Comment:
Seeing this here has me wondering: will forcing all to be strings end up making some of our downstream visualizations (in the app as well as in other external reporting) more complicated needing to know which of these values can | should be cast (back) to numbers?

Ultimately I suspect that it is a very small number of places where we will end up doing this (`cpu_count` being one), so maybe that's totally fine to deal with downstream. But seeing this here now made me think of the code I would need to write in order to order by `cpu_count` if I were using that in a report. 

I'm not trying to block this direction (I suspect like I say above, the number of places we do that will be small so it's worth it for the benefits of consistency), but it would be good to actively make that choice here :)

Reply:
You almost had me "yes OK let's allow for boolean in addition to string".

But I think it is a good choice when we think about the querying interface, and dynamic case querying first, esp. with browser UI support and URL construction:

- Imagine you put `=false` into a URL query string (and make it have a simple, obvious meaning).
- It's simpler to type `=false` into an input field than to fill in an actual boolean from a drop-down 

These two points above are important. I don't think we want to rely on a type-aware query language here.

It's a trade-off, as always. 

The mental burden of allowing for more than one type is actually rather tremendous whenever we build an interface for humans on the input and output side of things. Do we really want to pay attention to displaying the difference between a `false` and a `"false"` to the user?

More perspective: these case tags are primarily labels. Their role as metadata storage for an external system is secondary, in my opinion (it is not unimportant, but in a trade-off consideration some aspects by definition need to suffer a bit).

Let's think about why we are different from labels in the Prometheus ecosystem. There might be good arguments!

https://prometheus.io/docs/concepts/data_model/#metric-names-and-labels

> Every time series is uniquely identified by its metric name and optional key-value pairs called labels.

The similarity to benchmark name and case key/value pairs cannot be overlooked.

> Changing any label value, including adding or removing a label, will create a new time series.

Same for us.

> Labels enable Prometheus's dimensional data model: any given combination of labels for the same metric name identifies a particular dimensional instantiation of that metric (for example: all HTTP requests that used the method POST to the /api/tracks handler). The query language allows filtering and aggregation based on these dimensions. 

> A label with an empty label value is considered equivalent to a label that does not exist.

> Given a metric name and a set of labels, time series are frequently identified using this notation:
> `<metric name>{<label name>=<label value>, ...}`

IMO, stringified notation and strigified querying are to be valued considered before _external use_ of these labels.

Reply:
As labels, yes, all strings make sense. But tags don't start as labels, they start as arguments to benchmarks, which need to interpret them as types and work for authors (so switching to `memory_mapped="false"` is suboptimal). We can postprocess those arguments into strings for tags (like what happens for TPC-H query numbers -> IDs), but that indirection does come at a mental cost when tracing activity from benchmark through to frontend. Maybe we're ok with that cost?

Moreso than anything else, though, before enforcing that tags are strings we need PRs to all the benchmark repos to adjust tags such that lots of POSTs don't start failing. And we should clean up the docs which presently state `tags` is `Dict[str, Any]`.

Reply:
> I think you're saying that even those are used somewhere in code that we control, yes? Or is your feedback "no, I don't expect breakage of existing jobs when we start rejecting json objects and arrays as value types"?

Non-empty arrays are generated by [this benchmark](https://github.com/voltrondata-labs/arrowbench/blob/ebfab3d5535a1a7ba8680be4d0394296b9b1a5ee/R/bm-array-altrep-materialization.R#L19), but since we're not running that presently, we can edit it pretty easily before sending anything to Conbench. So asserting a flat structure shouldn't break anything active.

> > And we should clean up the docs which presently state tags is Dict[str, Any].
> 
> Of course. This PR tries to improve textual specification, did you see that? Which part of docs do you refer to with specifically "state tags is Dict[str, Any]"?

Importantly in benchadapt.BenchmarkResult: https://github.com/conbench/conbench/blob/16d0243833764f13f84978c43177871a89fb67a9/benchadapt/python/benchadapt/result.py#L43-L47 Less importantly in places like type hints in labs/benchmarks: https://github.com/voltrondata-labs/benchmarks/blob/75bc80e1bc75b428734f7a018a7f5460d6094274/benchmarks/_benchmark.py#L175-L177



> > [...] we might want to put together a little set of opinions on how to stringify, e.g. "2." vs "2.0", capitalization for booleans
> 
> I always try to point this out elsewhere: it's important to draw imaginary lines between areas of responsibility. These two statements certainly relate to the idea that benchmark authors _can_ implement a custom convention that allows for the information flow `any -> str/str -> any`, hopefully without ambiguity. We _can_ guide them to do so. But most importantly, we don't have to dictate a way for doing so.

My worry here is if we want types back later, e.g. for sorting or visualization. Maintaining a function like [type.convert()](https://rdrr.io/r/utils/type.convert.html) sounds like hell. If we're sure we're _only_ ever going to use these as labels, this doesn't matter.